server/src/main/java/org/elasticsearch/node/PluginComponentInterface.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.node;
+
+import org.elasticsearch.common.inject.Provider;
+import org.elasticsearch.plugins.ExtensiblePlugin;
+import org.elasticsearch.plugins.Plugin;
+
+/**
+ *<p>
+ * Describes the relationship between an interface and an implementation for a {@link Plugin} component. All {@link Plugin}'s
+ * provide a collection of components via the createComponents(...) that are made available for
+ * dependency injection. An {@link ExtensiblePlugin} may want to expose an interface such that extensions can change the implementation.
+ * Usage of this class allows the interface to made available for dependency injection where the implementation is defined at runtime.
+ *</p>
+ *     Usage:
+ * <pre>{@code
+public class MyPlugin extends Plugin implements ExtensiblePlugin
+{
+    @Override
+    public Collection<Object> createComponents(...) {
+        List<Object> components = new ArrayList<>();
+        components.add(new PluginComponentInterface<>(MyInterface.class, () -> this.myImplementation));
+        ...
+        return components;
+    }
+    @Override
+    public void loadExtensions(ExtensionLoader loader) {
+        List<MyInterface> myImplementations = loader.loadExtensions(MyInterface.class);
+        if(myImplementations.size() > 0) {
+            this.myImplementation = myImplementations.get(0); //use extension provided implementation
+        } else {
+            this.myImplementation = new StandardMyImplementation()
+        }
+    }
+...
+}
+public class TransportMyAction extends TransportMasterNodeAction<MyRequest, MyResponse> {
+    @Inject
+    public TransportMyAction(MyInterface myInterface) {
+        this.myInterface = myInterface; //implementation may vary depending on extensions defined for
+    }
+    ...
+}
+}</pre>
+ *<p>
+ * Note - usage of the class does not change any class loader isolation strategies nor visibility of the provided classes.
+ * </p>
+ * @param <I> The interface class
+ * @param <T> The implementation class
+ */
+public record PluginComponentInterface<I, T extends I> (Class<? extends I> clazz, Provider<T> provider) {}
```

Comment:
Why do we need a Provider here? Can we get away with capturing the actual instance? I thought the expectation would be the same: that we create such instance in createComponents? At worst, if there is a strong need for supported delayed creation, then can we use Supplier, ie not use our guice classes (so they can stay isolated to Node).

Reply:
nit on the name: "Interface" here is a bit confusing, since this is not an interface. Could we use something like PluginComponentBinding instead?

Reply:
> Why do we need a Provider here?

Good call.. it was in there for early approach which I abandoned and left it there since it could be useful.. however I didn't think about leaking guice classes. It's removed now [`e7ef710` (#94347)](https://github.com/elastic/elasticsearch/pull/94347/commits/e7ef7109c17540a9a4aaf7d9b637d6a98905b1d5).

Reply:
> Could we use something like PluginComponentBinding instead?

Done. [`e7ef710` (#94347)](https://github.com/elastic/elasticsearch/pull/94347/commits/e7ef7109c17540a9a4aaf7d9b637d6a98905b1d5)